### PR TITLE
Add end-to-end test and supervisor improvements

### PR DESCRIPTION
**Pull request by gigawhitlocks**
#### Summary

This branch is getting fairly significant so this is a checkpoint PR to start getting stuff reviewed and merged. The main feature on this branch is the end-to-end test, but there are also a number of improvements related to bugs found while building the test, which are necessary for the test to pass.

Done:
1) Adds an end-to-end test. The test creates a new Installation and imports a Slack workspace into it, and then verifies the users, team, and posts were properly imported. The test's runtime is around 10 minutes on average.
2) Fix bad synchronization logic when an Import is complete that we all saw and agonized about during the original merge of the Import supervisor into the AWAT. The logic is simpler now & with the additional testing I'm feeling much more confident about the new version than I ever was about the original. See commit 5581307
3) Revert a change I made in the last PR merged to master due to accidentally running tests against an old version of Mattermost (commit 163fcec)
4) Add cleanup: when a translation is done, clean up the local working directory. When an Import is done, delete the archive that was imported. The input zip that the operator uploads before starting is left in place. This is partially in commit 5581307 and partially in  4725c35 
5) 1-3, below

To Do (in this PR or after this PR, depending on when I get review):
~~1) Add a channel check to make sure channels were imported properly
2) Refactor the test into several tests so that progress can be displayed by go test and for clearer code
3) Clean up the tests with helper methods that are obviously missing~~
4) Test a Mattermost-format import
5) Test several imports fired off in parallel and make sure they don't interfere with each other, none of them are prematurely marked as `done`, and that they are processed in the right order
6) After https://github.com/mattermost/mattermost-server/pull/18084 is merged,  change [this line](https://github.com/mattermost/awat/compare/gigawhitlocks-e2e-testing-and-improvements?expand=1#diff-d5d12b6ea07b0a78c9ad1987e1486c565ecf16839f77a93e91b09475cf6c0e99R86) to point to a stable version of Mattermost Server

#### Ticket Link

This is at least part of https://mattermost.atlassian.net/browse/MM-37234

### CODE REVIEW

File: test/e2e/e2e_test.go [hunk 1]
```diff
@@ -0,0 +1,562 @@
...
+
+	params := &s3.ListObjectsV2Input{
+		Bucket: &bucketName,
+	}
+
+	p := s3.NewListObjectsV2Paginator(client, params,
+		func(o *s3.ListObjectsV2PaginatorOptions) {
+			if v := int32(1); v != 0 {
+				o.Limit = v
+			}
```
Comment (Szymongib) [inline test/e2e/e2e_test.go:538]: Not sure if I follow this part 🤔  
Isn't it essentially `o.Limit = int32(1)`?
Also is there a reason to set it to 1?

File: test/e2e/e2e_test.go [hunk 2]
```diff
@@ -0,0 +1,562 @@
...
+	checkPosts(t, env.provisioner, ci)
+}
+
+func setupEnv(t *testing.T) *environment {
+	t.Log("validate the environment and gather variables")
+
+	env, err := validatedEnvironment()
+	require.NoError(t, err)
+
+	err = ensureArtifactInBucket(env.bucket)
```
Comment (Szymongib) [inline test/e2e/e2e_test.go:81]: It seems like we call this function also in `validatedEnvironment`.
Btw does it make sense to check if the object exists in the bucket if we delete it anyway after the test? Maybe we can just call `PutObject` every time and overwrite the old one if exists?

**Review by Szymongib — COMMENTED** [review]
Great work on the test, I like detailed checks if everything is imported correctly. I think we can simplify some things and still might have some problems with edge cases of handling imports, but overall look good!

(Sorry if the review is premature but I will be out tomorrow and the day after so I did not want to block you.)

File: internal/supervisor/import.go [hunk 3]
```diff
@@ -88,16 +81,52 @@ func (s *ImportSupervisor) supervise() error {
...
+	case
+		// go ahead and mark Imports against Deleted Installations as
+		// complete
+		installation.State == cloud.InstallationStateDeleted:
+	case
+		// if the State of the Installation is stable and the Import was
+		// started at least a little while ago, we can safely assume that
+		// an import happened and was completed
+		installation.State == cloud.InstallationStateStable &&
+			i.StartAt < (model.Timestamp()-(time.Minute.Nanoseconds()/1000)):
```
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply internal/supervisor/import.go:127]: That edge case is addressed but it's a bit hard to reason about so I'll try to explain as best as I can.

When the lock fails, the Provisioner marks the Import complete and reports the error in the `Error` field, so the race condition doesn't occur. The Import supervisor comes back the next time it wakes up and finds the Import marked complete (with error!) and just moves on. 

The only way this race condition can occur is if the Provisioner succeeds in claiming the Import, but then fails to lock the Installation _and_ fails to talk back about the failure to lock to the AWAT -- in which case, the logic here will mark the Import done when it hasn't been performed, and it'll require manual reconciliant -- the Import has to be restarted manually. But that's the only downside.

To be clear: if the lock fails, the Import will require manual reconciliation _anyway_. In the original design, this is where the Autohealer would be running on another thread to eventually come along, look at Imports that failed with error, and possibly be able to automatically reconcile the state at a later time.

File: test/e2e/e2e_test.go [hunk 4]
```diff
@@ -0,0 +1,562 @@
...
+		&cloud.GetInstallationsRequest{
+			Paging:                      cloud.AllPagesNotDeleted(),
+			OwnerID:                     "awat-e2e",
+			IncludeGroupConfig:          false,
+			IncludeGroupConfigOverrides: false,
+		})
+	require.NoError(t, err)
+	// this is just a best-effort attempt to clean up from old test runs
+	// so just move on if it fails
+	if err == nil {
```
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:106]: We don't want to stop execution of the test here, which is what the comment is meant to communicate. I will make the comment more explicit.

File: test/e2e/users_test.go [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+// See LICENSE.txt for license information.
+//
+
+//+build e2e
+
+package e2e
+
+var correctUsers map[string]string
+
+func init() {
```
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply test/e2e/users_test.go:11]: Oh, no, haha, this is really stupid actually

I was fighting with gopls picking up these changes because of the build tag and I moved this into `init` from being a direct initialization because I was at that stage in debugging where you question your sanity and think "well maybe if I change this to the equivalent thing, it'll work, and then I can figure out why I thought the first thing was right in the first place?"

Anyway, then I fixed gopls and everything worked and I just forgot to change this to a direct initialization, especially since there's no real difference from a runtime perspective.

Thanks for listening to my story; I'll move this to a direct initialization :joy:

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:538]: Actually, you know, I copied the initialization of `p` from [an AWS example you can see here](https://github.com/aws/aws-sdk-go-v2/blob/main/example/service/s3/listObjects/listObjects.go#L58) and apparently didn't read very closely, because now that I am looking at it more carefully .... I don't know why it's done this way either? Maybe so I'll get caught copy-pasting code and not reading it closely enough :sweat_smile: 

What do you think? Should I just change it to `int32()` and assume the guy who wrote the AWS example was worried about different architectures that we don't support, or something?

File: test/e2e/e2e_test.go [hunk 6]
```diff
@@ -0,0 +1,562 @@
...
+	defer cancelFunc()
+
+	for p.HasMorePages() {
+		page, err := p.NextPage(ctx)
+		if err != nil {
+			return err
+		}
+
+		for _, item := range page.Contents {
+			if *item.Key == bucketName {
```
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:555]: Yeah this sure does look like a bug, thanks

File: test/e2e/e2e_test.go (hunk 2 above)
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:81]: Ah thanks for noticing the double-call, and yeah I think that does make sense, not sure I realized PutObject would overwrite (and testing AWS APIs can be a little bit laborious so please forgive me for not testing that assumption)

File: test/e2e/e2e_test.go (hunk 1 above)
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:538]: > Also is there a reason to set it to 1?

:shrug: it's what the docs said to do

File: internal/supervisor/import.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply internal/supervisor/import.go:127]: >If I am not mistaken Provisioner will never come back to this import as GetTranslationReadyToImport takes only imports with StartAt == 0. Installation was Stable the whole time tho, so after a minute pass, this will set it as complete even tho nothing has happened on the Provisioner side.

Yeah, this is actually intended, as I only wanted to stop a catastrophic or eternally-repeating error here, or a race condition when no lock is involved (as in the previous version of the code).

I have been reluctant to add an Installation state in the Provisioner of `import-complete` because conceptually it is the same as `stable` and because I was concerned about edge cases _there,_ but it makes this logic much simpler (basically just have the AWAT poll until the Installation state becomes `import-complete`, then have the AWAT issue a command to move the Installation to `stable` and mark the Import `import-complete` in its database) and the more I struggle with the myriad of possible corner cases in trying to navigate with logic _without_ that extra state, the more I'm thinking it's an error to try to continue without an Installation state of `import-complete`... let me just.. spike on that... hold tight, PR incoming

File: test/e2e/e2e_test.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply test/e2e/e2e_test.go:106]: Oh, crap, that `require` is not supposed to be there >.<

We don't want the text execution to stop, here.

File: internal/supervisor/import.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply internal/supervisor/import.go]: This is addressed with commit 0a81517 and PR https://github.com/mattermost/mattermost-cloud/pull/498

File: internal/supervisor/import.go [hunk 7]
```diff
@@ -88,16 +81,52 @@ func (s *ImportSupervisor) supervise() error {
...
+		if startedImportIsComplete(installation, i) {
+			i.CompleteAt = model.Timestamp()
 			err := s.store.UpdateImport(i)
 			if err != nil {
-				s.logger.WithError(err).Warnf("Import %s was done but couldn't be marked as such", i.ID)
+				s.logger.WithError(err).Warnf("Import %s was complete but couldn't be marked as such", i.ID)
 			}
+
+			key := fmt.Sprintf("%s.zip", i.TranslationID)
+			cfg, err := config.LoadDefaultConfig(context.TODO())
```
[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply internal/supervisor/import.go:92]: I took a second look at this and decided that this implementation is superior than initializing the client when the ImportSupervisor is initialized, because creating the AWS client can return error, so `NewImportSupervisor` would also need to return an error, and then that would need to be handled by the caller.

Since the ImportSupervisor is created only once, this would mean ending program execution upon failure to contact AWS at startup time, when the program isn't ready to do anything with AWS

With the current implementation, since the config is created when the Supervisor is ready to actually use it, the program won't produce an error unless it can't communicate with AWS at the time it actually needs to, and the existing mechanisms for retrial will work when AWS stops producing errors from `LoadDefaultConfig`

So actually, while I tend to pretty readily accept feedback, I'd like to leave this how it is unless you feel strongly that I am wrong to do so.

[the remark this answers is not in the document]
Comment (gigawhitlocks) [reply internal/supervisor/import.go:92]: Additionally, I've introduced v2 of the AWS SDK here and I don't want to:

1) get bogged down or sidetracked with replacing v1 of the AWS SDK elsewhere in the project so that this config can be reused, as I don't see this as a high-value or priority task

2) rewrite this stanza with v1 so that it uses a v1 AWS Client that can be used elsewhere in the application, because it seems silly to add new code w/ an SDK that is being deprecated

**Comment by gigawhitlocks** [comment]
This is ready for review @gabrieljackson

**Review by Szymongib — APPROVED** [review]
LGTM

**Review by gabrieljackson — APPROVED** [review]
Nice work!
